x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestGrantApiKeyAction.java

```diff
@@ -0,0 +1,128 @@
+        PARSER.declareString((req, str) -> req.getGrant().setUsername(str), new ParseField("username"));
+        PARSER.declareField((req, secStr) -> req.getGrant().setPassword(secStr), RestGrantApiKeyAction::getSecureString,
+            new ParseField("password"), ObjectParser.ValueType.STRING);
+        PARSER.declareField((req, secStr) -> req.getGrant().setAccessToken(secStr), RestGrantApiKeyAction::getSecureString,
+            new ParseField("access_token"), ObjectParser.ValueType.STRING);
+        PARSER.declareObject((req, api) -> req.setApiKeyRequest(api), (parser, ignore) -> CreateApiKeyRequestBuilder.parse(parser),
+            new ParseField("api_key"));
+    }
+
+    private static SecureString getSecureString(XContentParser parser) throws IOException {
+        return new SecureString(
+            Arrays.copyOfRange(parser.textCharacters(), parser.textOffset(), parser.textOffset() + parser.textLength()));
+    }
+
+    public RestGrantApiKeyAction(Settings settings, XPackLicenseState licenseState) {
+        super(settings, licenseState);
+    }
+
+    @Override
+    public List<Route> routes() {
+        return List.of(
+            new Route(POST, "/_security/api_key/grant"),
+            new Route(PUT, "/_security/api_key/grant"));
+    }
+
+    @Override
+    public String getName() {
+        return "xpack_security_grant_api_key";
+    }
+
+    @Override
+    protected RestChannelConsumer innerPrepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        String refresh = request.param("refresh");
+        try (XContentParser parser = request.contentParser()) {
+            final GrantApiKeyRequest grantRequest = PARSER.parse(parser, null);
+            if (refresh != null) {
+                grantRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.parse(refresh));
+            }
+            return channel -> client.execute(GrantApiKeyAction.INSTANCE, grantRequest, new ResponseListener(channel));
+        }
+    }
+
+    private class ResponseListener implements ActionListener<CreateApiKeyResponse> {
+        private final RestChannel channel;
+
+        ResponseListener(RestChannel channel) {
+            this.channel = channel;
+        }
+
+        @Override
+        public void onResponse(CreateApiKeyResponse response) {
+            try (XContentBuilder builder = channel.newBuilder()) {
+                channel.sendResponse(new BytesRestResponse(RestStatus.OK, response.toXContent(builder, channel.request())));
+            } catch (IOException e) {
+                sendFailure(e);
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            RestStatus status = ExceptionsHelper.status(e);
+            if (status == RestStatus.UNAUTHORIZED) {
+                sendFailure(new ElasticsearchSecurityException("Failed to authenticate api key grant", RestStatus.FORBIDDEN, e));
```

Comment:
I like this wrapping to translate 401 to 403!
Assuming this is the sole reason of existence for the inner class, in order to make it easier to read, I would make it more concise:
```
            return channel -> client.execute(GrantApiKeyAction.INSTANCE, grantRequest,
                    ActionListener.delegateResponse(new RestToXContentListener<>(channel),
                            (l, e) -> {
                                RestStatus status = ExceptionsHelper.status(e);
                                if (status == RestStatus.UNAUTHORIZED) {
                                    l.onFailure(new ElasticsearchSecurityException("Failed to authenticate api key grant",
                                            RestStatus.FORBIDDEN, e));
                                } else {
                                    l.onFailure(e);
                                }
                            }));
```

Reply:
Thanks @albertzaharovits that was a huge improvement.